### PR TITLE
Fixes #1725 Disable Card Preview Links.

### LINKED ISSUE #1346: Remove foo:bar from config
https://github.com/az-digital/az_quickstart/blob/caec8c5750e126107b4f14d981af0302ff57e722/modules/custom/az_paragraphs/az_paragraphs_cards/config/install/core.entity_view_display.paragraph.az_cards.default.yml#L17-L18

### LINKED ISSUE #1725: Clickable card previews are clickable
## Problem/Motivation
I don't want to go to the link of the card if I accidentally click on it from the paragraph preview while editing the page.

### Describe the bug
Paragraph preview should not render links.

### To Reproduce
1. Add clickable cards to a page
2. Collapse the paragraph
3. Observe clickable card functionality

### PR DESCRIPTION
**Pull request by tadean**
This PR makes card link previews not clickable. This is done by providing a new field formatter setting used by the card preview that avoids the default link behavior. The default card display has this setting turned off. The preview card setting has new configuration provided to provide its field formatter settings. An unneeded placeholder configuration setting is also removed.

This particular path (editing the preview display mode) was chosen because it involves no extra hooks and uses the built-in display system.

## Related issues
#1725 
Closes #1346
## How to test

- Create cards with links (esp clickable cards as these previews are most readily clickable)
- Verify links work on view display
- Verify links do not work in preview mode on edit page
- Verify pointer turns to a 'no' symbol on the edit page.
- Verify behavior can be successfully imported into new site
- Verify display settings page displays the correct "Interactive Links" setting for default and preview of the card paragraph.

### Arizona Quickstart (install profile, custom modules, custom theme)
- Patch release changes
   - [x] Bug fix
   - [x] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
- Minor release changes
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Non-critical brand change
   - [x] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- [ ] Other or unknown

### Drupal core
- Patch release changes
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major or minor level update
- [ ] Other or unknown

### Drupal contrib projects
- Patch release changes
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major level update
- [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by camikazegreen** [comment]
When using Firefox, I get a different hover indicator depending on whether or not the card has a link title. 

With link title: pointer
Without link title: not-allowed

All other behavior is as expected in both front-end and editing experience as both admin and content admin users.

File: modules/custom/az_card/src/Plugin/Field/FieldFormatter/AZCardDefaultFormatter.php [hunk 1]
```diff
@@ -73,11 +73,10 @@ public static function defaultSettings() {
...
 
-    // @todo Card style selection (based on custom config entities).
-    $element['foo'] = [
-      '#type' => 'textfield',
-      '#title' => $this->t('Foo'),
-      '#default_value' => $settings['foo'],
+    $element['interactive_links'] = [
+      '#type' => 'checkbox',
+      '#title' => $this->t('Interactive Links'),
+      '#default_value' => $settings['interactive_links'],
```
Comment (trackleft) [inline modules/custom/az_card/src/Plugin/Field/FieldFormatter/AZCardDefaultFormatter.php:79]: Possible description, though may be too wordy.
```suggestion
      '#default_value' => $settings['interactive_links'],
      '#description' => $this->t('If set, card links are clickable. Uncheck this setting to disable all card links. A common use-case is on the "Preview" view mode to prevent users from losing edit data if accidentally clicking on cards from the edit form. ');
```

File: modules/custom/az_card/src/Plugin/Field/FieldFormatter/AZCardDefaultFormatter.php [hunk 2]
```diff
@@ -73,11 +73,11 @@ public static function defaultSettings() {
...
+      '#description' => $this->t('If set, card links are clickable. Uncheck this setting to disable all card links. A common use-case is on the "Preview" view mode to prevent users from losing edit data if accidentally clicking on cards from the edit form.');
```
Comment (bberndt-uaz) [inline modules/custom/az_card/src/Plugin/Field/FieldFormatter/AZCardDefaultFormatter.php:80]: Suggestion: end line with a comma instead of semicolon to resolve PHPStan error

```suggestion
      '#description' => $this->t('If set, card links are clickable. Uncheck this setting to disable all card links. A common use-case is on the "Preview" view mode to prevent users from losing edit data if accidentally clicking on cards from the edit form.'),
```

**Review by joeparsons — DISMISSED** [review]

**Review by bberndt-uaz — COMMENTED** [review]
Tested in Probo using Chrome:
- Verify links work on view display
  - Verified
- Verify links do not work in preview mode on edit page
  - Links still work in Preview mode--they display a "Leave preview?" pop-up
- Verify pointer turns to a 'no' symbol on the edit page.
  - Links do not work on the Edit page, but as @camikazegreen noted, cards with text links have the not-allowed cursor but cards with buttons have the normal pointer cursor.

**Comment by joeparsons** [comment]
Follow-up tasks:

- Investigate and fix issue where cards become clickable again after being saved
- Investigate and fix issue with wrong cursor showing up (in Firefox?)

**Comment by bberndt-uaz** [comment]
Confirmed that the Edit page cursor issue is present in Chrome and Safari. Below, neither card is clickable but hovering over Card One shows the not-allowed cursor (due to the text link) while hovering over Card Two shows the normal pointer cursor for a link.

image

**Comment by tadean** [comment]
Added additional styles to cover button elements that are no-follow.

**Comment by tadean** [comment]
Converted no-follow js into a behavior so it is also applied to new and ajaxified content. (covers re-adding no-follow to collapsed decks)

**Review by bberndt-uaz — APPROVED** [review]
Tested in Probo and confirmed that on the Edit page for a node, when a card deck is collapsed, the links on all cards are not clickable and hovering over the links displays the not-allowed cursor. Both text links and button links correctly show the not-allowed cursor. Clicking the Edit button for a card deck (so it is no longer collapsed) makes the card links clickable with the standard pointer cursor. The node preview page behaves as before (the links are clickable and clicking them displays a confirmation dialog).

**Comment by trackleft** [comment]
Nice!

**Comment by tadean** [comment]
> Clicking the Edit button for a card deck (so it is no longer collapsed) makes the card links clickable with the standard pointer cursor. 

Added a fix for this in the card widget that the collapsed card widget also has the no-follow class. (this is the collapsed widget preview as compared to the paragraph preview)

**Review by trackleft — APPROVED** [review]

**Review by joeparsons — APPROVED** [review]
The recent changes all make sense and look good to me.

**Review by bberndt-uaz — APPROVED** [review]
